### PR TITLE
encrypt for invited users if history visibility allows.

### PR DESCRIPTION
**Pull request by ara4n**
fixes https://github.com/vector-im/riot-web/issues/2713

### CODE REVIEW

File: src/crypto/index.js [hunk 1]
```diff
@@ -1085,6 +1085,10 @@ Crypto.prototype._onRoomMembership = function(event, member, oldMembership) {
         console.log('Join event for ' + member.userId + ' in ' + roomId);
         // make sure we are tracking the deviceList for this user
         this._deviceList.startTrackingDeviceList(member.userId);
+    } else if (member.membership == 'invite' &&
+             this._clientStore.getRoom(roomId).shouldEncryptForInvitedUsers()) {
+        console.log('Invite event for ' + member.userId + ' in ' + roomId);
+        this._deviceList.startTrackingDeviceList(member.userId);
```
Comment (richvdh) [inline src/crypto/index.js]: Hard to check from my phone, but I worry about this causing (more) device tracking bugs, since we won't be reliably receiving device list updates from this user, but maybe we will think we are up to date, and hence not resync once the user does join the room?

Comment (dbkr) [reply src/crypto/index.js]: Agreed that encrypting messages to invited user could work unreliably if the server doesn't have the corresponding change, but it should be limited to invited users, right? Messages to joined user should work no worse than before?

Comment (dbkr) [reply src/crypto/index.js]: Actually no, I see what you mean. If the server doesn't have https://github.com/matrix-org/synapse/pull/3484 then it breaks when:

 1. Alice is in a room
 1. Bob is invited to the room. Alice starts tracking Bob's device list
 1. Bob adds a device. The server doesn't send an update poke.
 1. Bob joins. Alice's client will call `startTrackingDeviceList(bob)` but nothing will happen because Bob is already tracked.
 1. Alice sends a message, but she doesn't know about Bob's new device list.

In other words, I think this is perfectly safe in the situation where we can be sure the server will have https://github.com/matrix-org/synapse/pull/3484 . In the real world, one workaround could be to always force a re-fetch of the device list for a user when their membership transitions to `join` in any room?

Comment (ara4n) [reply src/crypto/index.js]: this of course assumes the synapse patch works, and is going to send enough devicelist updates for invited users correctly. i will test by seeing what happens if you add a device to an invited user before they join the room. (and then for the dumb synapse scenario, explicitly refetch the list when they join, as dave says)

Comment (ara4n) [reply src/crypto/index.js]: actually, it's going to be really hard for me to investigate this today, and we ideally need it for tomorrow. @dbkr: would you be able to check the concern above (and mitigate it by resyncing on /join if needed)?

Comment (dbkr) [reply src/crypto/index.js]: Following up, there was much discussion and the conclusion is that there's a big edge case here in that federated servers do not know what other servers are in the room at the time they've been invited. They therefore will not send device updates if a user logs in whilst their state is `invite`.

Comment (dbkr) [reply src/crypto/index.js]: Fixed. Having looks into forcing an update when a member joins the room, actually it's not necessary since synapse sends a device list update when the user joins the room anyway. Arguably this is synapse impl specific behaviour though.

**Comment by ara4n** [comment]
Conclusion here is that this works, other than for the edge case of https://github.com/matrix-org/synapse/issues/3504.

**Comment by ara4n** [comment]
...and doesn't cause any regressions; hence merging.
